### ISSUE TITLE
strange results on ShapeNet meshes

### ISSUE THREAD

**Issue by lodurality** [MClemot/SkeletonLearning#1]
Hi, great work!

You skeletons look very impressive in the paper and I want to try them for my project but I am getting weird results on ShapeNet. See attached image: your skeleton is the rightmost; leftmost object is original shape; second skeleton from the left is Mean Curvature Skeleton. 

![Selection_999(1011)](https://github.com/MClemot/SkeletonLearning/assets/8215811/f98577ce-cedd-424a-be6e-ae4d06cb2772)

Link to data: https://drive.google.com/drive/u/0/folders/11LP-2r7cLvet3_Ou-yOX2s67wz5h7Ijk

11194326770b1225d49b2e04785f8492.off -- input mesh
skel_learning_skeleton.obj -- output of main.py with delta=0.02

I am almost certain that this probably can be fixed with proper choice of hyperparameters but I have a hard time figuring this out. 

Would you kindly help me with this?

**Comment by MClemot** [comment]
Hi, thank you for your interest in our work!

With these kinds of shapes with very thin structures, we found out it was better to use a scaling (e.g. 1.5) instead of an offset for the inner balls dilatation in the set cover formulation (see Section 6.3 of Coverage Axis, Dou et al.). This is what I got using a 1.5 factor. I also had to reduce the weight of the TV regularization because in this case it tends to damage thin structures. I updated the repo to have an easy control on the said factor.

Note that with default parameters the trained network fails to distinguish the two wires, with a messy SDF estimation inside the estimated wire, which explains the inaccurate skeleton in this area. I guess it could be fixed with a proper choice of hyperparameters of the network and its training.
![lamp](https://github.com/MClemot/SkeletonLearning/assets/82047826/25ae586a-504d-4fed-aad2-417287d82fc2)
```
import torch
import sys
from io3d import to_obj
from geometry import sample_mesh
from build_neural_skeleton import build_neural_skeleton

device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')

str_input = sys.argv[1]
str_output = sys.argv[2]

pc, nc, M = sample_mesh(str_input, 100000)
vertices, edges, triangles, net, _, _ = build_neural_skeleton(pc, nc, scaling=1.5, lambda_tv=1)
to_obj(vertices * M, str_output, lines=edges, tri=triangles)
```

**Comment by lodurality** [comment]
Thank you for the fast reply, Mattéo!

I am having trouble reproducing your result with script that you gave to me. Image below is what I am getting with snippet you provided below -- results seem worse than main.py. 

I thought that maybe is the issue is that I am using older version of pytorch but I have also run your code in clean environment using your GitHub instructions and I am getting same results. 

Couple of additional notes/questions:
+ The meshes that I am working with are simplified watertight meshes produced by mesh-fusion algorithm that adds significant amount noise and often messes up normals. This might explain bad SDF estimation. As I understand, you worked with Thingi10K meshes that has much cleaner meshes. Do you think this can be fixed with proper hyperparameter choice?

+ It seems that your meshes scaled in [-1,1]^3 range while my meshes are scaled in [-0.5, 0.5]^3. Does it mean I should retrain your networks with pretrain.py? It seems that I should replace the line `pts_random = torch.rand((batch_size, 3), device = device)*2-1` with `pts_random = torch.rand((batch_size, 3), device = device)*1-0.5` in my case. Do you think this would help?

+ The significant appeal of your skeletons to me is that your algorithm outputs simplicial graph. Is there a way to increase number of vertices in it, so it gives smoother approximation of 1D/2D parts of the skeleton? I wasn't able to figure out the way to control this. 

+ Ideally, I want settings that preserve topology (connectivity and homotopy group) for the final skeleton even it becomes more complicated. I have added another example to[ google drive](https://drive.google.com/drive/u/0/folders/11LP-2r7cLvet3_Ou-yOX2s67wz5h7Ijk) (112058b3f8c94177d2dba55bc7acf0bc.off) and ideally I want to preserve chain links. Current output that I am getting is below.  

Again, thank you for the great work and I really appreciate you taking your time to answer my questions!

![Selection_999(1012)](https://github.com/MClemot/SkeletonLearning/assets/8215811/9efde4a3-d93c-40cf-af81-bd4b31dc2a7d)
![Selection_999(1013)](https://github.com/MClemot/SkeletonLearning/assets/8215811/c69fb65d-a4a8-436a-9869-0c42612bf17c)

**Comment by MClemot** [comment]
Hi Dmitry,
That's strange, just to be sure, did you use the last version of the repo ? The `scaling` option did nothing in the previous version, I updated that.
To answer your questions:
* My interpretation is that (reasonably) bad normals can degrade the SDF estimation near the surface but not too much far from the surface, which is what interests us the most to extract the skeleton. However it is completely possible to reduce the weight of the normal alignment loss (in front of the `sdf_loss_align` call in `nn.py`) if you consider they are too bad. In the two examples you sent to me, I am not convinced it is necessary to do so.
* You shouldn't have to do anything for normalization, we normalize automatically the input into [-1,1]^3 and scale back the output to the original scale.
* Yes! you can control the "level of detail" / smoothness of the skeleton with the delta parameter, or the factor if you use a scaling-based set cover. The smaller the delta is, the more points the skeleton will feature. Have a look at Section 6.4 / Figure 14 in [Coverage Axis, Dou et al.], or the example below with the bitore.obj example from the Objects folder, with delta = 0.15 on the left and delta = 0.01 on the right.
![bitore_example_delta](https://github.com/MClemot/SkeletonLearning/assets/82047826/41b9fb0d-a60d-4630-ad01-60595461bdd4)
* This is what I obtained with the same script as in my previous answer. Note that this method comes with no guarantees about the conservation of the homotopy type. In this case it seems that the SDF optimization step really struggles with the chain and doesn't achieve to discriminate the links, at least with the default hyperparameters.
![Screenshot from 2023-10-15 17-16-00](https://github.com/MClemot/SkeletonLearning/assets/82047826/f2c1ba66-7ce2-42d6-b59e-592d58a267af)

**Comment by lodurality** [comment]
Hi Mattéo!

I am sorry for the late reply -- I working on deadline and forgot to reply right away. 

I ended up using my own skeletons for my project but I really appreciate that found time to reply and clarify my questions. 

Best of luck with your research!

Sincerely,
Dmitry
